chain/network/src/peer_manager.rs

```diff
@@ -556,6 +580,79 @@ impl PeerManagerActor {
         });
     }
 
+    /// Select one peer and send signal to stop connection to it gracefully.
+    /// Selection process:
+    ///     Create a safe set of peers, and among the remaining peers select one at random.
+    ///     If the number of outbound connections is less or equal than minimum_outbound_connections,
+    ///         add all outbound connections to the safe set.
+    ///     While the length of the safe set is less than safe_set_size:
+    ///         Among all the peers we have received a message within the last peer_recent_time_window,
+    ///             find the one we connected earlier and add it to the safe set.
```

Comment:
What if a peer connected and then went offline? Why do we put it in the safe set?

Reply:
Do you mean, regarding outbound connections? Notice that otherwise we are checking for peers we have received a message recently, so no way it is offline.

Regarding outbound connections, it is preferred to keep them since it means we really have a very low number of outbound connections. eventually (sooner than later) number of outbound connections will normalize and if dropping connection is needed, they will be eligible to be dropped.

Reply:
In case the node is really offline, in less than a minute we will notice that and it will be outside of the safe set by then, so it is not really a big problem.